File: userland/src/libc.rs
extern crate alloc;
use core::alloc::{GlobalAlloc, Layout};
use core::arch::asm;

pub struct ProcessAllocator {}

impl ProcessAllocator {
    pub unsafe fn init(&mut self, heap_start: usize, heap_size: usize) {}
}

unsafe impl GlobalAlloc for ProcessAllocator {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        malloc(layout.size()) as *mut u8
    }

    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {
        todo!();
    }
}

#[global_allocator]
static ALLOCATOR: ProcessAllocator = ProcessAllocator {};

pub fn getpid() -> u64 {
    let mut _pid = 0xdeadbeef;

    unsafe {
        asm!("
            push rdi
            mov rdi, 2

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            pop rdi
            ",
            out("rax") _pid,
            options(nostack)
        );
    }

    return _pid;
}

pub fn write(filedescriptor: i64, payload: *const u64, len: usize) {
    unsafe {
        asm!("
            push rdi
            mov rdi, 1

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            
            pop rdi
        ",
            in("r8") filedescriptor,
            in("r9") payload as u64,
            in("r10") len,
            options(nostack),
            clobber_abi("C")
        );
    }
}

pub fn draw_pixel(x: u32, y: u32, color: u8) {
    unsafe {
        asm!("
            push rdi
            mov rdi, 3

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            
            pop rdi
        ",
            in("r8") x,
            in("r9") y,
            in("r10") color as u64,
            options(nostack),
            clobber_abi("C")
        );
    }
}

pub fn malloc(size: usize) -> u64 {
    let mut address: u64;

    unsafe {
        asm!("
            push rdi
            mov rdi, 4

            push r11
            push rcx

            syscall

            pop rcx
            pop r11

            pop rdi
        ",
            in("r8") size,
            out("rax") address,
            options(nostack),
            clobber_abi("C")
        );
    }

    return address;
}

pub fn fopen() {
    unsafe {
        asm!(
            "
            push rdi
            mov rdi, 5

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            pop rdi
            ",
            options(nostack)
        );
    }
}

pub fn fclose() {
    // TODO does nothing for now
}

pub fn fwrite() {
    // TODO does nothing for now
}

pub fn fseek(offset: usize, origin: usize) {
    unsafe {
        asm!(
            "
            push rdi
            mov rdi, 7

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            pop rdi
            ",
            options(nostack),
            in("r8") offset,
            in("r9") origin,
        );
    }
}

pub fn feof() -> u64 {
    let mut eof: u64;

    unsafe {
        asm!(
            "
            push rdi
            mov rdi, 9

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            pop rdi
            ",
            options(nostack),
            out("rax") eof,
        );
    }

    return eof;
}

pub fn ftell() -> u64 {
    let mut position: u64;

    unsafe {
        asm!(
            "
            push rdi
            mov rdi, 8

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            pop rdi
            ",
            options(nostack),
            out("rax") position,
        );
    }

    return position;
}

//size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
pub fn fread(ptr: *mut u8, size: usize, nmemb: usize) {
    unsafe {
        asm!(
            "
            push rdi
            mov rdi, 6

            push r11
            push rcx
        
            syscall
        
            pop rcx
            pop r11
            pop rdi
            ",
            options(nostack),
            in("r8") ptr,
            in("r9") size*nmemb,
        );
    }
}

pub struct Printer {}

impl core::fmt::Write for Printer {
    fn write_str(&mut self, s: &str) -> core::fmt::Result {
        write(1, s.as_bytes().as_ptr() as *const u64, s.len());
        Ok(())
    }
}

#[macro_export]
macro_rules! printf {
    () => {    };
    ($($arg:tt)*) => {{
        let mut printer = crate::libc::Printer {};
        core::fmt::write(&mut printer, core::format_args!($($arg)*)).unwrap();
    }};
}

File: userland/src/libc.c
#include "libc.h"
#include <inttypes.h> 
#include <stdbool.h>
#include <stdint.h>

uint64_t strlen( const char* str ) {
    int len = 0;

    while (str[len] != '\0') {
        len++;
    }

    return len;
}

bool strcmp(const char* a, const char* b) {
    int i =0;

    while (a[i] != '\0') {
        if (a[i]==b[i]) {
            i++;
            continue;
        } else {
            return false;
        }
    }

    return true;
}

// Write function using syscall
void write(uint64_t filedescriptor, const char* payload, uint64_t len) {
    uint64_t result;
    DO_SYSCALL(1, result, filedescriptor, (uintptr_t)payload, len);
}

// Get process ID
uint64_t getpid() {
    uint64_t pid;
    DO_SYSCALL(2, pid, 0, 0, 0);  // No additional arguments needed
    return pid;
}

// Draw a pixel on screen
void draw_pixel(uint32_t x, uint32_t y, uint8_t color) {
    uint64_t result;
    DO_SYSCALL(3, result, x, y, color);
}

// Allocate memory
void* malloc(int size) {
    uint64_t address;
    DO_SYSCALL(4, address, size, 0, 0);  // Only size is passed
    return (void*)address;
}

// Free memory (currently does nothing)
void free(void* address) {
    // TODO: Implement the free function
}

// Open a file
void* fopen(const char* filename, const char* options) {
    if (!strcmp(filename, "devdatadoom1.wad")) {
        return 0;
    }

    write(1, "fopen: ", strlen("fopen: "));
    write(1, filename, strlen(filename));

    uint64_t handle;
    DO_SYSCALL(5, handle, (uintptr_t)filename, (uintptr_t)options, 0);
    return (void*)handle;
}

// Close a file (currently does nothing)
void fclose(void* handle) {
    // TODO: Implement the fclose function
}

// Write to a file (currently does nothing)
int fwrite(void* handle, const void* foo, int bar) {
    write(1, "fwrite: ", strlen("fwrite: "));
    return 0;  // TODO: Implement fwrite
}

// Seek within a file
int fseek(void* handle, int offset, doom_seek_t origin) {
    uint64_t result;
    DO_SYSCALL(7, result, offset, origin, 0);
    return result;
}

// Check if end of file
int feof(void* handle) {
    uint64_t eof;
    DO_SYSCALL(9, eof, (uintptr_t)handle, 0, 0);
    return eof;
}

// Get the current position in a file
int ftell(void* handle) {
    uint64_t position;
    DO_SYSCALL(8, position, (uintptr_t)handle, 0, 0);
    return position;
}

// Read from a file
int fread(void* handle, void* ptr, int size) {
    uint64_t read_bytes;
    DO_SYSCALL(6, read_bytes, (uintptr_t)ptr, size, 0);
    return read_bytes;
}

// Draw the framebuffer
uint64_t draw_framebuffer(const uint8_t* framebuffer) {
    uint64_t result;
    DO_SYSCALL(10, result, (uintptr_t)framebuffer, 0, 0);
    return result;
}

// Switch VGA mode
uint64_t switch_vga_mode(bool vga_on) {
    uint64_t result;
    DO_SYSCALL(11, result, vga_on, 0, 0);
    return result;
}

// Get the state of a key
bool get_keystate(int key) {
    uint64_t state;
    DO_SYSCALL(12, state, key, 0, 0);
    return (bool)state;
}

void get_time(int* sec, int* usec) {
    uint64_t result;
    DO_SYSCALL(13, result, sec, usec, 0);
}